// lavas-core-vue/dist/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
/**
 * @file constants.js
 * @author lavas
 */

const LAVAS_DIRNAME_IN_DIST = exports.LAVAS_DIRNAME_IN_DIST = 'lavas';
const SKELETON_DIRNAME = exports.SKELETON_DIRNAME = 'skeletons';
const ASSETS_DIRNAME_IN_DIST = exports.ASSETS_DIRNAME_IN_DIST = 'static';
const LAVAS_CONFIG_FILE = exports.LAVAS_CONFIG_FILE = 'lavas.config.js';

const CLIENT_MANIFEST = exports.CLIENT_MANIFEST = 'vue-ssr-client-manifest.json';
const SERVER_BUNDLE = exports.SERVER_BUNDLE = 'vue-ssr-server-bundle.json';
const TEMPLATE_HTML = exports.TEMPLATE_HTML = 'index.html.tmpl';
const SPA_TEMPLATE_HTML = exports.SPA_TEMPLATE_HTML = 'spa.html.tmpl';
const SSR_TEMPLATE_HTML = exports.SSR_TEMPLATE_HTML = 'ssr.html.tmpl';
const CONFIG_FILE = exports.CONFIG_FILE = 'config.json';
const DEFAULT_ENTRY_NAME = exports.DEFAULT_ENTRY_NAME = 'index';
const STORE_FILE = exports.STORE_FILE = 'store.js';
const DEFAULT_SKELETON_PATH = exports.DEFAULT_SKELETON_PATH = '@/core/Skeleton.vue';
const BUILD_SCRIPT = exports.BUILD_SCRIPT = 'build.js';
