// lavas-core-vue/dist/utils/loader.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

exports.vueLoaders = vueLoaders;
exports.cssLoaders = cssLoaders;
exports.styleLoaders = styleLoaders;

var _extractTextWebpackPlugin = require('extract-text-webpack-plugin');

var _extractTextWebpackPlugin2 = _interopRequireDefault(_extractTextWebpackPlugin);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function vueLoaders({ cssSourceMap, cssMinimize, cssExtract, babelOptions }) {
    return {
        loaders: (0, _assign2.default)({
            js: {
                loader: 'babel-loader',
                options: babelOptions
            }
        }, cssLoaders({
            cssSourceMap,
            cssMinimize,
            cssExtract
        }))
    };
} /**
   * @file utils.loader.js
   * @author lavas
   */
function cssLoaders({ cssMinimize, cssSourceMap, cssExtract }) {

    const cssLoader = {
        loader: 'css-loader',
        options: {
            minimize: cssMinimize,
            sourceMap: cssSourceMap
        }
    };

    const postcssLoader = {
        loader: 'postcss-loader',
        options: {
            sourceMap: cssSourceMap
        }
    };

    // generate loader string to be used with extract text plugin
    function generateLoaders(loader, loaderOptions) {
        let loaders = [cssLoader, postcssLoader];

        if (loader) {
            loaders.push({
                loader: loader + '-loader',
                options: (0, _assign2.default)({}, loaderOptions, {
                    sourceMap: cssSourceMap
                })
            });
        }

        // Extract CSS when that option is specified
        // (which is the case during production build)
        if (cssExtract) {
            return _extractTextWebpackPlugin2.default.extract({
                use: loaders,
                fallback: 'vue-style-loader'
            });
        }

        return ['vue-style-loader', ...loaders];
    }

    // https://vue-loader.vuejs.org/en/configurations/extract-css.html
    return {
        css: generateLoaders(),
        postcss: generateLoaders(),
        less: generateLoaders('less'),
        sass: generateLoaders('sass', {
            indentedSyntax: true
        }),
        scss: generateLoaders('sass'),
        stylus: generateLoaders('stylus'),
        styl: generateLoaders('stylus')
    };
}

// Generate loaders for standalone style files (outside of .vue)
function styleLoaders(options) {
    let output = [];
    let loaders = cssLoaders(options);

    (0, _keys2.default)(loaders).forEach(extension => {
        output.push({
            test: new RegExp(`\\.${extension}$`),
            use: loaders[extension],
            name: extension
        });
    });

    return output;
}
