// lavas-core-vue/dist/utils/logger.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _chalk = require('chalk');

var _chalk2 = _interopRequireDefault(_chalk);

var _ora = require('ora');

var _ora2 = _interopRequireDefault(_ora);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @file Logger
 * @author panyuqi
 * @desc use ora to show log messages
 */

const LOG = {
    DEBUG: {
        LEVEL: 3,
        COLOR: _chalk2.default.green
    },
    INFO: {
        LEVEL: 2,
        COLOR: _chalk2.default.green
    },
    WARN: {
        LEVEL: 1,
        COLOR: _chalk2.default.yellow
    },
    ERROR: {
        LEVEL: 0,
        COLOR: _chalk2.default.red
    }
};
const DEFAULT_OPTIONS = {
    level: LOG.INFO.LEVEL,
    textColor: _chalk2.default.white
};

class Logger {
    constructor(options) {
        this.options = options;
        this.spinner = null;
    }

    log(tag, content, chalkFunc, update, done) {
        let chalkedContent = chalkFunc.bold(`[Lavas ${tag}] `) + this.options.textColor(content);
        if (update) {
            if (!this.spinner) {
                this.spinner = (0, _ora2.default)(chalkedContent).start();
            } else {
                this.spinner.succeed(chalkedContent);
                this.spinner = null;
            }
        } else {
            console.log(chalkedContent);
        }
    }

    debug(tag, content, update, done) {
        if (this.options.level < LOG.DEBUG.LEVEL) {
            return;
        }
        this.log(tag, content, LOG.DEBUG.COLOR, update, done);
    }

    info(tag, content, update, done) {
        if (this.options.level < LOG.INFO.LEVEL) {
            return;
        }
        this.log(tag, content, LOG.INFO.COLOR, update, done);
    }

    warn(tag, content, update, done) {
        if (this.options.level < LOG.WARN.LEVEL) {
            return;
        }
        this.log(tag, content, LOG.WARN.COLOR, update, done);
    }

    error(tag, content, update, done) {
        if (this.options.level < LOG.ERROR.LEVEL) {
            return;
        }
        this.log(tag, content, LOG.ERROR.COLOR, update, done);
    }
}

exports.default = new Logger(DEFAULT_OPTIONS);
module.exports = exports['default'];
